### PR TITLE
Re-use pre-converted kernel arguments when launching kernels.

### PR DESCRIPTION
**Pull request by maleadt**
Addresses part of https://github.com/JuliaGPU/CUDA.jl/issues/2456, by avoiding a redundant `cudaconvert` call on kernel launches. The idea is that, instead of first calling `cudaconvert` + `typeof` to drive compilation, and then for the kernel's arguments when doing the actual launch (see below for why this split is necessary), we only call `cudaconvert` at launch time when there's a mismatch between an argument's type and what the kernel was compiled for. This makes it possible for `@cuda` to forward the already-converted arguments to the kernel, avoiding a secondary conversion at launch time.

Note that the whole distinction between compile-time and launch-time `cudaconvert` exists to make it possible to do the following:

```julia
k = @cuda launch=false kernel(args)
k(args)
```

i.e., to make `launch=false` user-friendly and let users pass non-converted arguments into a pre-compiled kernel, the API _has_ to call `cudaconvert` twice, once when compiling as part of `@cuda launch=false`, and once when calling the kernel. The alternative would be to have users explicitly call `cudaconvert(args)` when launching the kernel, but I didn't want to expose the implementation detail that arguments are converted.

cc @vchuravy 

@pxl-th @tgymnich This pattern probably applies to other GPU back-ends as well. CUDA.jl's `cudaconvert` is somewhat expensive because of our automatic memory tracking (https://juliagpu.org/post/2024-05-28-cuda_5.4/#tracked_memory_allocations, https://github.com/JuliaGPU/CUDA.jl/blob/76e2972814a0e7910f35ed3ad17b1a9198628f34/src/memory.jl#L494-L598)

### CODE REVIEW

File: src/compiler/execution.jl [hunk 1]
```diff
@@ -238,15 +245,29 @@ function Base.show(io::IO, ::MIME"text/plain", k::AbstractKernel{F,TT}) where {F
     print(io, "CUDA.$(nameof(typeof(k))) for $(k.f)($(join(TT.parameters, ", ")))")
 end
 
-@inline @generated function call(kernel::AbstractKernel{F,TT}, args...; call_kwargs...) where {F,TT}
+@inline @generated function (kernel::AbstractKernel{F,TT})(args...; call_kwargs...) where {F,TT}
```
Comment (wsmoses) [inline src/compiler/execution.jl]: can this be made vararg like it was previously? x/ref https://github.com/EnzymeAD/Enzyme.jl/issues/1745

Not sure if thats the fix, but it feels like it

Comment (maleadt) [reply src/compiler/execution.jl]: The call is still vararg.

Can you adapt the extension to handle this version of kernel calls?

Comment (wsmoses) [reply src/compiler/execution.jl]: sorry by vararg I mean `args::Vararg{Any,N}`

Comment (maleadt) [reply src/compiler/execution.jl]: Same error.

File: src/compiler/execution.jl [hunk 2]
```diff
@@ -111,7 +111,7 @@ macro cuda(ex...)
                     $kernel_tt = Tuple{map(Core.Typeof, $kernel_args)...}
                     $kernel = $cufunction($kernel_f, $kernel_tt; $(compiler_kwargs...))
                     if $launch
-                        $kernel($(var_exprs...); $(call_kwargs...))
+                        $kernel($kernel_args...; $(call_kwargs...))
```
Comment (wsmoses) [inline src/compiler/execution.jl]: Okay now I understand the issue, @maleadt previously we didn't need to differentiate through the cudaconvert to the kernel, but now we do and those haven't been added.

I think this (or something similar) ought resolve: https://github.com/JuliaGPU/CUDA.jl/pull/2476

**Comment by maleadt** [comment]
I wonder if alternatively, to avoid the questionable conditional call to `cudaconvert`, we should just use a typevar in the AbstractKernel indicating whether arguments ought to be converted or not (only setting that to `false` when re-using the args in the context of `@cuda`-generated code).

**Pull request merged**